include/exadg/fluid_structure_interaction/driver_precice_fluid.h

```diff
@@ -0,0 +1,594 @@
+      auto                    exadg_terminal_ale =
+        std::make_shared<ExaDG::preCICE::InterfaceCoupling<dim, dim, Number>>();
+
+      // Poisson mesh movement
+      if(this->application->get_parameters_fluid().mesh_movement_type ==
+         IncNS::MeshMovementType::Poisson)
+      {
+        std::vector<unsigned int> quad_indices;
+        if(this->application->get_parameters_ale_poisson().spatial_discretization ==
+           Poisson::SpatialDiscretization::DG)
+          quad_indices.emplace_back(ale_poisson_operator->get_quad_index());
+        else if(this->application->get_parameters_ale_poisson().spatial_discretization ==
+                Poisson::SpatialDiscretization::CG)
+          quad_indices.emplace_back(ale_poisson_operator->get_quad_index_gauss_lobatto());
+        else
+          AssertThrow(false, ExcNotImplemented());
+
+        quadrature_point_locations = exadg_terminal_ale->setup(
+          ale_matrix_free,
+          ale_poisson_operator->get_dof_index(),
+          quad_indices,
+          this->application->get_boundary_descriptor_ale_poisson()->dirichlet_cached_bc);
+      }
+      // Elasticity mesh movement
+      else if(this->application->get_parameters_fluid().mesh_movement_type ==
+              IncNS::MeshMovementType::Elasticity)
+      {
+        std::vector<unsigned int> quad_indices;
+        quad_indices.emplace_back(ale_elasticity_operator->get_quad_index_gauss_lobatto());
+
+        quadrature_point_locations = exadg_terminal_ale->setup(
+          ale_matrix_free,
+          ale_elasticity_operator->get_dof_index(),
+          quad_indices,
+          this->application->get_boundary_descriptor_ale_elasticity()->dirichlet_cached_bc);
+      }
+      else
+      {
+        AssertThrow(false, ExcNotImplemented());
+      }
+      this->precice->add_read_surface(quadrature_point_locations,
+                                      ale_matrix_free,
+                                      exadg_terminal_ale,
+                                      this->precice_parameters.ale_mesh_name,
+                                      {this->precice_parameters.displacement_data_name});
+    }
+
+    // structure to fluid
+    {
+      std::vector<unsigned int> quad_indices;
+      quad_indices.emplace_back(fluid_operator->get_quad_index_velocity_linear());
+      quad_indices.emplace_back(fluid_operator->get_quad_index_velocity_nonlinear());
+      quad_indices.emplace_back(fluid_operator->get_quad_index_velocity_gauss_lobatto());
+
+      VectorType velocity_structure;
+      fluid_operator->initialize_vector_velocity(velocity_structure);
+      auto exadg_terminal_fluid =
+        std::make_shared<ExaDG::preCICE::InterfaceCoupling<dim, dim, Number>>();
+      auto quadrature_point_locations = exadg_terminal_fluid->setup(
+        fluid_matrix_free,
+        fluid_operator->get_dof_index_velocity(),
+        quad_indices,
+        this->application->get_boundary_descriptor_fluid()->velocity->dirichlet_cached_bc);
```

Comment:
The same as in comment https://github.com/exadg/exadg/pull/148/files#r803717139

Reply:
@DavidSCN did you resolve this?

Reply:
No, I will add TODO comments in the code regarding this.

Reply:
Done